feature/text_based.py

```diff
@@ -0,0 +1,724 @@
+
+        Parameters
+        ----------
+        input_df: pd.DataFrame
+            Input data frame with categories and features of content to select from.
+
+        categories: List[int]
+            List of columns in data that contains categories/labels to be covered.
+
+        featurization_method: TextWiser
+
+        method: str, default="exact"
+            Method used to perform content selection. Supported options are:
+            - "exact" maximizes content diversity and minimizes the number of content to cover all content labels.
+            - "kmeans" clusters the content into selection_size number of clusters and then selects the items closest
+              to the centroid of each of the clusters. This method does not consider the content categories/labels.
+            - "greedy" performs greedy heuristic selecting items with max unit coverage until all items are covered
+            - "random" performs a random selection.
+
+        cost_metric: str, default = None
+
+        Returns
+        -------
+        List of indices in data that are selected.
+        """
+
+        # Process data
+        self.cost_metric = cost_metric
+        self.featurization_method = featurization_method
+        self._process_df(input_df, categories, method, cost_metric)
+
+        # Run multi-level set covering optimization
+        if method == "exact":
+            selected = self._select_multi_level_optimization()
+        # Run KMeans selection algorithm
+        elif method == "kmeans":
+            selected = self._select_kmeans()
+        # Run greedy heuristic
+        elif method == "greedy":
+            selected = self._select_greedy()
+        # Run random selection algorithm
+        elif method == "random":
+            selected = self._select_random()
+        else:
+            raise NotImplementedError(f"{method} has not been implemented.")
+
+        # Sort selection
+        selected.sort()
+
+        return selected
+
+    def _process_df(self, input_df: pd.DataFrame, categories: List[int], method: str, cost_metric: Optional[str]) \
+            -> NoReturn:
+        """
+
+        Parameters
+        ----------
+        input_df: pd.DataFrame
+            Input data frame with categories and features of content to select from.
+
+        categories: List[int]
+            List of columns in data that contains categories/labels to be covered.
+
+        method: str
+            The function need method to prevent performing featurization in
+             the case that optimization method is random(t or unicost) or greedy(unicost)
+
+        cost_metric: str
+            The function need cost metric to prevent performing featurization in
+             the case that cost metric is None or unicost (for random and greedy)
+
+        Returns
+        -------
+        No return
+       --------
+        * categories is the list of labels. code for labels with categories moved to utils.py
```

Comment:
not following this line with the asterix *

Reply:
I modified the text to: "We have implemented two versions of this function. In the version in the text_based.py, the categories parameter actually refers to the labels (rather than categories) to be covered. The version of code for generating labels from categories has been moved to utils.py."

In the _process_df function in utils.py, each label is created by concatenating multiple category values from the input dataframe, which means each label can have different categories. The categories for each label are defined by the categories parameter, which is a list of column names in the input dataframe that contain the category values.

 The _process_df function is currently used, we assume that categories is a numpy array of labels, and not a list of columns in the input data frame containing the labels. Therefore, if we assume that we already have a matrix of labels and there are no categories.